### PR TITLE
Add UPDATE_NVRAM in /etc/sysconfig/bootloader (bsc#1157550).

### PR DESCRIPTION
**Pull request by hramrach**
Signed-off-by: Michal Suchanek <msuchanek@suse.de>

### CODE REVIEW

File: src/lib/bootloader/grub2base.rb [hunk 1]
```diff
@@ -400,6 +408,18 @@ def trusted_boot_summary
           "<a href=\"enable_trusted_boot\">(" + _("enable") + ")</a>"
         end
     end
+
+    # Update nvram shown in summary screen
+    #
+    # @return [String]
+    def update_nvram_summary
```
Comment (jreidinger) [inline src/lib/bootloader/grub2base.rb:415]: is this summary used anywhere?

File: src/lib/bootloader/grub2_widgets.rb [hunk 2]
```diff
@@ -423,6 +423,42 @@ def validate
...
+
+    def init
+      self.value = grub2.update_nvram
+    end
+
+    def store
+      grub2.update_nvram = value
+    end
+
+    def validate
```
Comment (jreidinger) [inline src/lib/bootloader/grub2_widgets.rb:457]: this method is not needed. by default it true.

File: src/lib/bootloader/grub2_widgets.rb [hunk 3]
```diff
@@ -423,6 +423,42 @@ def validate
...
+      _("Update &NVRAM Entry")
+    end
+
+    def help
+      res = _("<p><b>Update NVRAM</b> will add nvram entry for the bootloader\n" \
+          "in the firmware.\n" \
+          "This is usually desirable unless you want to preserve specific settings\n" \
+          "or need to work around firmware issues.\n" \
+          "\n")
+      res += "</p>"
```
Comment (jreidinger) [inline src/lib/bootloader/grub2_widgets.rb:444]: three notes to this help text:

1. having `<p>` in localized text and `</p>` not is strange and can confuse translators
2. it is richtext so `\n` is not needed at all
3. `Update NVRAM` does not match `Update NVRAM Entry`. bold text should mention widget name.

File: src/lib/bootloader/grub2.rb [hunk 4]
```diff
@@ -107,6 +107,7 @@ def summary(simple_mode: false)
 
       result << secure_boot_summary if Systeminfo.secure_boot_available?(name)
       result << trusted_boot_summary if Systeminfo.trusted_boot_available?(name)
+      result << update_nvram_summary if Systeminfo.nvram_available?(name)
```
Comment (jreidinger) [inline src/lib/bootloader/grub2.rb:110]: well, I start feeling that it is a bit abusing of this shortcuts. Original idea was to have in bootloader summary the most common option that user want just change by one click. I think this is quite expert option and should not be mentioned in summary page.

File: src/lib/bootloader/autoyast_converter.rb [hunk 5]
```diff
@@ -235,7 +236,8 @@ def export_grub2efi(res, bootloader)
       # only for grub2, not for others
       GRUB2_BOOLEAN_MAPPING = {
         "secure_boot"  => :secure_boot,
-        "trusted_grub" => :trusted_boot
+        "trusted_grub" => :trusted_boot,
+        "update_nvram"  => :update_nvram,
```
Comment (jreidinger) [inline src/lib/bootloader/autoyast_converter.rb:240]: so update nvram can be done also for non-EFI systems?

File: src/lib/bootloader/grub2_widgets.rb (hunk 2 above)
Comment (hramrach) [reply src/lib/bootloader/grub2_widgets.rb:457]: ok

File: src/lib/bootloader/grub2_widgets.rb (hunk 3 above)
Comment (hramrach) [reply src/lib/bootloader/grub2_widgets.rb:444]: The texts commonly include \n on end of line.

File: src/lib/bootloader/grub2base.rb (hunk 1 above)
Comment (hramrach) [reply src/lib/bootloader/grub2base.rb:415]: src/lib/bootloader/grub2.rb:      result << secure_boot_summary if Systeminfo.secure_boot_available?(name)
src/lib/bootloader/grub2.rb:      result << trusted_boot_summary if Systeminfo.trusted_boot_available?(name)
src/lib/bootloader/grub2.rb:      result << update_nvram_summary if Systeminfo.nvram_available?(name)
src/lib/bootloader/grub2.rb:      result << url_location_summary unless no_location
src/lib/bootloader/grub2.rb:      order_sum = disk_order_summary
src/lib/bootloader/grub2.rb:    def disk_order_summary
src/lib/bootloader/grub2.rb:    def url_location_summary
src/lib/bootloader/grub2base.rb:    def secure_boot_summary
src/lib/bootloader/grub2base.rb:    def trusted_boot_summary
src/lib/bootloader/grub2base.rb:    def update_nvram_summary
src/lib/bootloader/grub2efi.rb:      result << secure_boot_summary if Systeminfo.secure_boot_available?(name)
src/lib/bootloader/grub2efi.rb:      result << trusted_boot_summary if Systeminfo.trusted_boot_available?(name)
src/lib/bootloader/grub2efi.rb:      result << update_nvram_summary if Systeminfo.nvram_available?(name)

File: src/lib/bootloader/autoyast_converter.rb (hunk 5 above)
Comment (hramrach) [reply src/lib/bootloader/autoyast_converter.rb:240]: it applies to EFI and PPC

File: src/lib/bootloader/grub2.rb (hunk 4 above)
Comment (hramrach) [reply src/lib/bootloader/grub2.rb:110]: that's certainly open to discussion.

**Comment by hramrach** [comment]
Why are commas at end of hash rejected?
Lack of the final comma makes updating the value list error-prone and introduces needless code churn adding/removing commas to otherwise untouched lines.

**Comment by jreidinger** [comment]
@hramrach well, we just follow upstream style guide and deviate only with good reason. If you are interested in discussion you can see https://github.com/rubocop-hq/ruby-style-guide/issues/76

**Comment by hramrach** [comment]
> @hramrach well, we just follow upstream style guide and deviate only with good reason. If you are interested in discussion you can see [rubocop-hq/ruby-style-guide#76](https://github.com/rubocop-hq/ruby-style-guide/issues/76)

What upstream style guide?

The discussion points out the reasons already given - missing final comma in multiline list makes it harder to edit code and review patches. It also discusses at length that either style is preferred by different people because it makes the code more readable for them personally.

Rubocop has settings for either.

**Comment by hramrach** [comment]
Added separate patch for removing --no-nvram from the ARM workaround so that it does not conflict with UPDATE_NVRAM

**Comment by hramrach** [comment]
The second install on the removable system does not happen - or is tested wrong.

**Comment by jreidinger** [comment]
implemented in https://github.com/yast/yast-bootloader/pull/617

**Pull request closed without merging**